Rework Config Watcher and configuration reload

pkosiec opened the pull request:
<!-- Thank you for your contribution. Before you submit the pull request:
1. Follow contributing guidelines, templates, the recommended Git workflow, and any related documentation.
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

## Description

Changes proposed in this pull request:

- ...

## Limitations



## To do

- [ ] Documentation - as a follow-up PR after merging this one

## Related issue(s)

Resolves #765

huseyinbabal commented:
@pkosiec I was testing locally, I suggest to use following notation for botkube image to prevent apple m1 problem, plus your pr already generates docker image manifest for multiple archs. 
```
image:
  repository: kubeshop/pr/botkube
  tag: 774-PR
```
I have apple m1 problem with config watcher, will fix and rerun the tests
<img width="896" alt="Screen Shot 2022-09-29 at 10 44 35" src="https://user-images.githubusercontent.com/1237982/192971259-6f040b42-5a09-49cc-8621-01e9bdd9f65d.png">

pkosiec commented:
Oh sorry @huseyinbabal - I pushed my own image as there were problem with the CI build. But yeah, it was amd64 only. 

Now as I rebased the PR, please use the values:
```
image:
  repository: kubeshop/pr/botkube
  tag: 774-PR
```

huseyinbabal commented:
> Oh sorry @huseyinbabal - I pushed my own image as there were problem with the CI build. But yeah, it was amd64 only.
> 
> Now as I rebased the PR, please use the values you posted 🙂 Will update PR desc as well. Thanks!

sure, I am also building k8s-sidecar for arm64, will push to ghcr
<img width="948" alt="Screen Shot 2022-09-29 at 10 52 44" src="https://user-images.githubusercontent.com/1237982/192973223-62a42a03-e35e-4b65-a1bc-2026651fe580.png">

pkosiec commented:
Ah, yes, I forgot about making k8s-sidecar multiarch build 😞 I'm sorry. Original k8s-sidecar is multiarch indeed, but again, I built amd64 version only... Time to switch from Intel to M1/M2 as well 😄 Thank you Huseyin for help and sorry for troubles!

huseyinbabal commented:
Tested with Socket and Socket Slack mode, it works well 👍 
<img width="741" alt="Screen Shot 2022-09-29 at 11 22 45" src="https://user-images.githubusercontent.com/1237982/192979874-284d55af-4bc0-401a-93d4-f85fbc0e6287.png">
<img width="742" alt="Screen Shot 2022-09-29 at 11 23 00" src="https://user-images.githubusercontent.com/1237982/192979883-e5a7e7e7-56ba-4229-9560-0c684b46e019.png">

mszostok commented on pkg/bot/teams.go:
```diff
@@ -553,5 +553,9 @@ func replaceEmojiTagsWithActualOne(content string) string {
 
 // emojiMapping holds mapping between emoji tags and actual ones.
```
please do me a favour and fix bug on `main`

Constructor:
```
	longFormatter := interactive.NewMDFormatter(longLineFormatter, interactive.MdHeaderFormatter)
	shortFormatter := interactive.NewMDFormatter(shortLineFormatter, interactive.MdHeaderFormatter)

	return &Teams{
		log:             log,
		executorFactory: executorFactory,
		reporter:        reporter,
		botName:         cfg.BotName,
		ClusterName:     clusterName,
		AppID:           cfg.AppID,
		AppPassword:     cfg.AppPassword,
		Notification:    cfg.Notification,
		bindings:        cfg.Bindings,
		commGroupName:   commGroupName,
		MessagePath:     msgPath,
		Port:            port,
		conversations:   make(map[string]conversation),
		botMentionRegex: botMentionRegex,
		longFormatter:   longFormatter,
		shortFormatter:  shortFormatter,
	}, nil
``

New funcs instead of `mdLineFormatter`:
```go
// longLineFormatter represents new line formatting for MS Teams where message has multiple sections.
// Unfortunately, it's different from all others integrations.
func longLineFormatter(msg string) string {
	// e.g. `:rocket:` is not supported by MS Teams, so we need to replace it with actual emoji
	msg = replaceEmojiTagsWithActualOne(msg)
	return fmt.Sprintf("%s<br>", msg)
}

func shortLineFormatter(msg string) string {
	// e.g. `:rocket:` is not supported by MS Teams, so we need to replace it with actual emoji
	msg = replaceEmojiTagsWithActualOne(msg)
	return fmt.Sprintf("%s\n", msg)
}
```

Updated `convertInteractiveMessage` func:
```go
func (b *Teams) convertInteractiveMessage(in interactive.Message) string {
	if in.HasSections() {
		// MS Teams doesn't respect multiple new lines, so it needs to be rendered
		// with `<br>` tags instead  ¯\_(ツ)_/¯
		return interactive.MessageToMarkdown(b.longFormatter, in)
	}
	return interactive.MessageToMarkdown(b.shortFormatter, in)
}
```

the bug was unfortunately introduced here: https://github.com/kubeshop/botkube/pull/737/files#diff-dab5cea33bef23767be4be070914501c91d73cab9748905293abb4bc47b7b97c

pkosiec replied:
Sure thing, we'll do that 👍

pkosiec commented:
Hey @huseyinbabal
- I changed the defaults as agreed 🙂 
- I repushed the image to have it multiarch: https://github.com/orgs/kubeshop/packages/container/k8s-sidecar/43648586?tag=ignore-initial-events. I also updated the PR desc with command I used for this: https://github.com/kubeshop/k8s-sidecar/pull/1
- I also put @mszostok's fix for MS Teams formatting. Mateusz already tested the changes he posted.

Hope we'll be ready to merge this soon 🚀

huseyinbabal commented:
> 

I retested with final changes, and it works 🚀
